lib/cosmos/tools/cmd_tlm_server/gui/replay_tab.rb

```diff
@@ -0,0 +1,242 @@
+      @op_button_layout.addWidget(@reverse_play)
+      @stop = Qt::PushButton.new(Cosmos.get_icon('stop-26.png'), '')
+      @stop.connect(SIGNAL('clicked()')) { CmdTlmServer.replay_backend.stop() }
+      @op_button_layout.addWidget(@stop)
+      @play = Qt::PushButton.new(Cosmos.get_icon('play-26.png'), '')
+      @play.connect(SIGNAL('clicked()')) { CmdTlmServer.replay_backend.play() }
+      @op_button_layout.addWidget(@play)
+      @step_forward = Qt::PushButton.new(Cosmos.get_icon('fast_forward-26.png'), '')
+      @step_forward_timer = Qt::Timer.new
+      @step_forward_timeout = 100
+      @step_forward_timer.connect(SIGNAL('timeout()')) { CmdTlmServer.replay_backend.step_forward(); @step_forward_timeout = (@step_forward_timeout / 2).to_i; @step_forward_timer.start(@step_forward_timeout) }
+      @step_forward.connect(SIGNAL('pressed()')) { CmdTlmServer.replay_backend.step_forward(); @step_forward_timeout = 300; @step_forward_timer.start(@step_forward_timeout) }
+      @step_forward.connect(SIGNAL('released()')) { @step_forward_timer.stop }
+      @op_button_layout.addWidget(@step_forward)
+      @move_end = Qt::PushButton.new(Cosmos.get_icon('end-26.png'), '')
+      @move_end.connect(SIGNAL('clicked()')) { CmdTlmServer.replay_backend.move_end() }
+      @op_button_layout.addWidget(@move_end)
+      @op_layout.addLayout(@op_button_layout)
+
+      # Speed Selection
+      @playback_delay = nil
+      @speed_select = Qt::ComboBox.new
+      @variants = []
+      @variants << [Qt::Variant.new(nil), nil]
+      @speed_select.addItem("No Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.001), 0.001]
+      @speed_select.addItem("1ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.002), 0.002]
+      @speed_select.addItem("2ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.005), 0.005]
+      @speed_select.addItem("5ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.01), 0.01]
+      @speed_select.addItem("10ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.05), 0.05]
+      @speed_select.addItem("50ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.125), 0.125]
+      @speed_select.addItem("125ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.25), 0.25]
+      @speed_select.addItem("250ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(0.5), 0.5]
+      @speed_select.addItem("500ms Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(1.0), 1.0]
+      @speed_select.addItem("1s Delay", @variants[-1][0])
+      @variants << [Qt::Variant.new(-1.0), -1.0]
+      @speed_select.addItem("Realtime", @variants[-1][0])
+      @speed_select.setMaxVisibleItems(11)
+      @speed_select.connect(SIGNAL('currentIndexChanged(int)')) do
+        CmdTlmServer.replay_backend.set_playback_delay(@speed_select.itemData(@speed_select.currentIndex).value)
+      end
+      @speed_layout = Qt::FormLayout.new()
```

Comment:
This comment applies throughout: do all these need to be instance variables? This one (for example) is only used here.

Reply:
In older versions of qtbindings they did need to be instance variables to prevent getting garbage collected.  In newer versions it isn't necessary but doesn't hurt anything.

Reply:
Except for my eyes!